eth-riscv-runtime/src/lib.rs

```diff
@@ -116,6 +120,29 @@ pub fn msg_sender() -> Address {
     Address::from_slice(&bytes)
 }
 
+pub fn msg_value() -> U256 {
+    let first: u64;
+    let second: u64;
+    let third: u64;
+    let fourth: u64;
+    unsafe {
+        asm!("ecall", lateout("a0") first, lateout("a1") second, lateout("a2") third, lateout("a3") fourth, in("t0") u8::from(Syscall::CALLVALUE));
+    }
+    U256::from_limbs([first, second, third, fourth])
+}
+
+pub fn msg_sig() -> [u8; 4] {
+    let sig = unsafe { slice_from_raw_parts(CALLDATA_ADDRESS + 8, 4) };
+    sig.try_into().unwrap()
+}
+
+pub fn msg_data() -> Vec<u8> {
+    let length = unsafe { slice_from_raw_parts(CALLDATA_ADDRESS, 8) };
+    let length = u64::from_le_bytes([length[0], length[1], length[2], length[3], length[4], length[5], length[6], length[7]]) as usize;
+    let calldata = unsafe { slice_from_raw_parts(CALLDATA_ADDRESS + 8, length) };
+    calldata.to_vec()
```

Comment:
we could return a slice to avoid allocation?

Reply:
Ah yes!